qiskit_experiments/visualization/plotters/base_plotter.py

```diff
@@ -0,0 +1,453 @@
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""Base plotter abstract class"""
+
+import warnings
+from abc import ABC, abstractmethod
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+from qiskit_experiments.framework import Options
+from qiskit_experiments.visualization.drawers import BaseDrawer
+from qiskit_experiments.visualization.style import PlotStyle
+
+
+class BasePlotter(ABC):
+    """An abstract class for the serializable figure plotters.
+
+    A plotter takes data from an experiment analysis class or experiment and plots a given figure using a
+    drawing backend. Sub-classes define the kind of figure created and the expected data.
+
+    Data is split into series and figure data. Series data is grouped by series name (str). For
+    :class:`CurveAnalysis`, this is the model name for a curve fit. For series data associated with a
+    single series name and figure data, data-values are identified by a data-key (str). Different data
+    per series and figure must have a different data-key to avoid overwriting values. Experiment and
+    analysis results can be passed to the plotter so appropriate graphics can be drawn on the figure
+    canvas. Series data is added to the plotter using :meth:`set_series_data` whereas figure data is
+    added using :meth:`set_figure_data`. Series and figure data are retrieved using :meth:`data_for` and
+    :attr:`figure_data` respectively.
+
+    Series data contains values to be plotted on a canvas, such that the data can be grouped into subsets
+    identified by their series name. Series names can be thought of as legend labels for the plotted
+    data, and as curve names for a curve-fit. Figure data is not associated with a series or curve and is
+    instead only associated with the figure. Examples include analysis reports or other text that is
+    drawn onto the figure canvas.
+
+    # TODO: Add example usage and description of options and plot-options.
+    """
+
+    def __init__(self, drawer: BaseDrawer):
+        """Create a new plotter instance.
+
+        Args:
+            drawer: The drawer to use when creating the figure.
+        """
+        # Data to be plotted, such as scatter points, interpolated fits, and confidence intervals
+        self._series_data: Dict[str, Dict[str, Any]] = {}
+        # Data for figure-wide drawing, unrelated to series data, such as text or fit reports.
+        self._figure_data: Dict[str, Any] = {}
+
+        # Options for the plotter
+        self._options = self._default_options()
+        # Plotter options that have changed, for serialization.
+        self._set_options = set()
+
+        # Plot options that are updated in the drawer when `plotter.figure()` is called
+        self._plot_options = self._default_plot_options()
+        # Plot options that have changed, for serialization.
+        self._set_plot_options = set()
```

Comment:
Can you add a comment, either here or in the class description above, why we have `_options` and `_plot_options`? Something to explain when an option should be in one or the other variable? The current comments hint at this but need to be more explicit.

Reply:
Addressed in 8c588dc.